rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,1081 @@
+have no resource-level expiration (though that is allowed), are meant to have
+their spec modified over time by users or automation tools, and publish
+information about their current state in the immutable (to users) `status`
+field.
+
+``` yaml
+kind: token
+version: v2
+metadata:
+  name: my-join-token
+spec:
+  bot_name: example
+
+  join_method: bound-keypair
+  bound_keypair:
+    # `onboarding` parameters control initial join behavior
+    onboarding:
+      # If set, no joining secret is generated; the secret exchange ceremony is
+      # skipped and instance will directly prove its identity using its private
+      # key. It is an error for a public key to be associated with more than one
+      # token, and creation or update will fail if a public key is reused.
+      # May not be modified after resource creation. Note that public keys may
+      # be rotated, so refer to `.status.bound_keypair.bound_public_key` for the
+      # currently bound key information.
+      # This key should be written in SSH public key format, including the
+      # algorithm.
+      initial_public_key: null
+
+      # If set, use an explicit initial joining secret; if both this and
+      # `public_key` are unset, a value will be generated server-side and
+      # stored in `.status.bound_keypair.initial_join_secret`
+      initial_join_secret: ""
+
+      # Use of `initial_join_secret` must take place before this timestamp. May
+      # be modified if initial secret has not yet been consumed.
+      must_join_before: "2025-03-01T21:45:40.104524Z"
+
+    # Parameters to tune joining behavior, both on first join and when rejoining
+    # when the regular identity expires.
+    joining:
+      # If true, `total_joins` is ignored and bots may rejoin indefinitely;
+      # must be opt-in.
+      unlimited: false
+
+      # Total number of allowed joins; this may be incremented to allow
+      # additional rejoins, even if a bot identity has already expired. May
+      # be decremented, but only by the current value of
+      # `.status.bound_keypair.remaining_joins`. This value must be at least 1
+      # for a bot to perform an initial join.
+      total_joins: 10
+
+      # If set, joining is only valid before this timestamp; may be
+      # incremented or reset to the empty string to allow rejoining once
+      # expired.
+      may_join_until: "2026-03-01T21:45:40.104524Z"
+
+      # Insecure is an optional flag that enables insecure joining and
+      # rejoining. This method disables generation counter checks during joining
+      # and rejoining. When combined with the `unlimited` flag, this allows
+      # unlimited reuse of this token provided the client has access to the
+      # keypair. This may be useful in certain cases - like use in unsupported
+      # CI/CD providers - but cannot offer the same security assurances and
+      # should be used with care.
+      insecure: false
```

Comment:
Small update: this is something we've mulled over for a while and was loosely mentioned in a section below ("explicitly insecure token joining"). I think we can accomplish a lightweight version of that with just this additional flag; users would enable both `unlimited` and `insecure`, and could then share a keypair in e.g. a CI/CD secret.

We've had a number of customers - especially those using unsupported CI/CD platforms - looking to take some advantage of Machine ID. I've added it now since I think now (while I'm WIP on the initial implementation of this RFD) is the most efficient time to add the flag, but am happy to cut this if there's any concern.